go/vt/vttablet/onlineddl/executor.go

```diff
@@ -0,0 +1,320 @@
+		hash: ShortRandomHash(),
+		env:  env,
+		pool: connpool.NewPool(env, "ExecutorPool", tabletenv.ConnPoolConfig{
+			Size:               1,
+			IdleTimeoutSeconds: env.Config().OltpReadPool.IdleTimeoutSeconds,
+		}),
+		tabletTypeFunc: tabletTypeFunc,
+		ts:             ts,
+		ticks:          timer.NewTimer(migrationCheckInterval),
+	}
+}
+
+func (e *Executor) initSchema(ctx context.Context) error {
+	defer e.env.LogError()
+
+	conn, err := e.pool.Get(ctx)
+	if err != nil {
+		return err
+	}
+	defer conn.Recycle()
+	_, err = withDDL.Exec(ctx, sqlValidationQuery, conn.Exec)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// InitDBConfig initializes keysapce
+func (e *Executor) InitDBConfig(keyspace, shard string) {
+	e.keyspace = keyspace
+	e.shard = shard
+}
+
+// Open opens database pool and initializes the schema
+func (e *Executor) Open() error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.isOpen {
+		return nil
+	}
+	e.pool.Open(e.env.Config().DB.AppWithDB(), e.env.Config().DB.DbaWithDB(), e.env.Config().DB.AppDebugWithDB())
+	e.ticks.Start(e.checkMigrations)
+	e.isOpen = true
+
+	return nil
+}
+
+// Close frees resources
+func (e *Executor) Close() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if !e.isOpen {
+		return
+	}
+
+	e.ticks.Stop()
+	e.pool.Close()
+	e.isOpen = false
+}
+
+func (e *Executor) ghostPanicFlagFileName() string {
+	return fmt.Sprintf("/tmp/ghost.%s.panic.flag", e.hash)
+}
+
+// readMySQLVariables contacts the backend MySQL server to read some of its configuration
+func (e *Executor) readMySQLVariables(ctx context.Context) (host string, port int, readOnly bool, err error) {
+	conn, err := e.pool.Get(ctx)
+	if err != nil {
+		return host, port, readOnly, err
+	}
+	defer conn.Recycle()
+
+	tm, err := conn.Exec(ctx, "select @@global.hostname, @@global.port, @@global.read_only", 1, false)
+	if err != nil {
+		return host, port, readOnly, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "could not read MySQL variables: %v", err)
+	}
+	if len(tm.Rows) != 1 {
+		return host, port, readOnly, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "unexpected result for MySQL variables: %+v", tm.Rows)
+	}
+	host = tm.Rows[0][0].ToString()
+	if p, err := evalengine.ToInt64(tm.Rows[0][1]); err != nil {
+		return host, port, readOnly, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "could not parse @@global.port %v: %v", tm, err)
+	} else {
+		port = int(p)
+	}
+	if rd, err := evalengine.ToInt64(tm.Rows[0][2]); err != nil {
+		return host, port, readOnly, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "could not parse @@global.read_only %v: %v", tm, err)
+	} else {
+		readOnly = (rd != 0)
+	}
+	return host, port, readOnly, nil
+}
+
+// Execute validates and runs a gh-ost process.
+// Validation included testing the backend MySQL server and the gh-ost binray itself
+// Execution runs first a dry run, then an actual migration
+func (e *Executor) Execute(ctx context.Context, target querypb.Target, alias topodatapb.TabletAlias, schema, table, alter string) error {
```

Comment:
Do we also want to check schema consistency (temp gh-ost tables), ongoing migrations (temp gh-ost flag files), and maybe replication health between master and relicas?

Reply:
- replication health: definitely; `gh-ost` uses throttling, and we will want to supply `gh-ost` with the identity of a replica it will check for lag. Also, if we ever deploy/implement `freno` in Vitess, then `gh-ost` will use that as a standard throttling mechanism.

- ongoing migrations: some work already in, possibly not pushed; as mentioned above we will avoid concurrent migrations and only allow one migration ta a time.

- schema consistency: explain?